core-processor/src/processor.rs

```diff
@@ -42,33 +49,197 @@ enum SuccessfulDispatchResultKind {
     Success,
 }
 
-/// Process program & dispatch for it and return journal for updates.
-pub fn process<A: ProcessorExt + EnvExt + IntoExtInfo + 'static, E: Environment<A>>(
+/// Checked parameters for message execution across processing runs.
+pub struct PreparedMessageExecutionContext {
+    gas_counter: GasCounter,
+    gas_allowance_counter: GasAllowanceCounter,
+    dispatch: IncomingDispatch,
+    origin: ProgramId,
+    balance: u128,
+    program: Program,
+    memory_size: WasmPageNumber,
+}
+
+impl PreparedMessageExecutionContext {
+    /// Returns reference to the GasCounter.
+    pub fn gas_counter(&self) -> &GasCounter {
+        &self.gas_counter
+    }
+}
+
+/// Defines result variants of the function `prepare`.
+#[allow(clippy::large_enum_variant)]
```

Comment:
instead of allowing clippy, can we use box for Ok case?

Reply:
Yep, we are able to, but do we really want to have unnecessary heap allocation?

Reply:
Discussed